From e853d6468bec6986fd917f44684efb49ce17293e Mon Sep 17 00:00:00 2001
Message-Id: <e853d6468bec6986fd917f44684efb49ce17293e.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Kausal Malladi <Kausal.Malladi@intel.com>
Date: Sat, 31 Jan 2015 16:44:20 +0530
Subject: [PATCH 80/95] MUST_REBASE [VPG]: video/adf: CHV Brightness Contrast
 correction

This patch implements core functions that programs
Brightness and Contrast correction registers on CHV platform.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-4167
Change-Id: Ib58ad3d3875ec3eb45d61ec893a0bd9fb27404e7
Signed-off-by: Kausal Malladi <Kausal.Malladi@intel.com>
---
 .../adf/intel/core/vlv/chv_color_correction.c      |   86 ++++++++++++++++++++
 1 file changed, 86 insertions(+)

diff --git a/drivers/video/adf/intel/core/vlv/chv_color_correction.c b/drivers/video/adf/intel/core/vlv/chv_color_correction.c
index aad45fa..32c105f 100644
--- a/drivers/video/adf/intel/core/vlv/chv_color_correction.c
+++ b/drivers/video/adf/intel/core/vlv/chv_color_correction.c
@@ -77,6 +77,8 @@ const struct color_property chv_plane_color_corrections[] = {
 		.prop_id = contrast,
 		.len = CHV_CB_VALS,
 		.name = "contrast",
+		.set_property = chv_set_contrast,
+		.disable_property = chv_disable_contrast,
 		.validate = chv_validate,
 	},
 
@@ -85,6 +87,8 @@ const struct color_property chv_plane_color_corrections[] = {
 		.prop_id = brightness,
 		.len = CHV_CB_VALS,
 		.name = "brightness",
+		.set_property = chv_set_brightness,
+		.disable_property = chv_disable_brightness,
 		.validate = chv_validate,
 	},
 
@@ -154,6 +158,88 @@ const u16 deGamma_LUT_B[CHV_DEGAMMA_VALS] = {
 	14181, 14723, 15276, 15842, 16383
 };
 
+bool chv_set_brightness(struct color_property *property, u64 *data, u8 plane_id)
+{
+	u32 val, new_val, reg, sprite;
+
+	/* If sprite plane enabled */
+	sprite = plane_id;
+	if (!(REG_READ(CHV_CLRMGR_SPCNTR(sprite)) &
+			SP_ENABLE)) {
+		pr_err("ADF: CM: Sprite plane %d not enabled\n", sprite);
+		return false;
+	}
+
+	reg = CHV_CLRMGR_SPCB(sprite);
+
+	/* Clear current values */
+	val = REG_READ(reg) & ~(CHV_BRIGHTNESS_MASK);
+
+	/*Get new values */
+	new_val = *data & CHV_BRIGHTNESS_MASK;
+
+	/* Update */
+	val |= new_val;
+	REG_WRITE(reg, val);
+	property->lut[0] = new_val;
+
+	/* Set status */
+	if (new_val == CHV_BRIGHTNESS_DEFAULT)
+		property->status = false;
+	else
+		property->status = true;
+	pr_info("ADF: CM: Set Brightness correction to %d successful\n",
+				new_val);
+	return true;
+}
+
+bool chv_disable_brightness(struct color_property *property, u8 plane_id)
+{
+	u64 data = CHV_BRIGHTNESS_DEFAULT;
+	return chv_set_brightness(property, &data, plane_id);
+}
+
+bool chv_set_contrast(struct color_property *property, u64 *data, u8 plane_id)
+{
+	u32 val, new_val, reg, sprite;
+
+	/* If sprite plane enabled */
+	sprite = plane_id;
+	if (!(REG_READ(CHV_CLRMGR_SPCNTR(sprite)) &
+			SP_ENABLE)) {
+		pr_err("ADF: CM: Sprite plane %d not enabled\n", sprite);
+		return false;
+	}
+
+	reg = CHV_CLRMGR_SPCB(sprite);
+
+	/* Clear current value. Contrast correction position is bit [26:18] */
+	val = REG_READ(reg) &
+		~(CHV_CONTRAST_MASK << CHV_CONTRAST_SHIFT);
+
+	/* Get new value */
+	new_val = *data & CHV_CONTRAST_MASK;
+
+	/* Update */
+	val |= (new_val << CHV_CONTRAST_SHIFT);
+	REG_WRITE(reg, val);
+	property->lut[0] = new_val;
+
+	/* Set status */
+	if (new_val == CHV_CONTRAST_DEFAULT)
+		property->status = false;
+	else
+		property->status = true;
+	pr_info("ADF CM: Set Contrast to 0x%x successful\n", new_val);
+	return true;
+}
+
+bool chv_disable_contrast(struct color_property *property, u8 plane_id)
+{
+	u64 data = CHV_CONTRAST_DEFAULT;
+	return chv_set_contrast(property, &data, plane_id);
+}
+
 /* Core function to enable degamma block */
 bool chv_set_degamma(struct color_property *property, u64 *data, u8 pipe_id)
 {
-- 
1.7.9.5

